src/main/java/com/google/sps/storage/DatastoreGuides.java

```diff
@@ -61,4 +63,8 @@ public Guide put(Guide guide){
     public List<Guide> queryAll(){
         return listFromQueryResults(datastore.run(query));
     }
+
+    public Guide get(long id){
+        return toGuide(datastore.get(keyFactory.newKey(id)));
```

Comment:
You could check the return value from datastore.get and return null. However, this just means you have more null-checking to do up the stack, and it's easy to forget. There are a couple options I would prefer:
 - Return a `java.util.Optional<Guide>` - this makes it clear that you need to check whether the value is present
 - Throw an exception if the entity doesn't exist. Define a NotFoundException type (subclassing RuntimeException) and then use an error handler to map it to an HTTP 404. Subclassing RuntimeException (or another error class) is preferable to using it directly because that way you can catch and handle it in your code if needed.

Reply:
Since the `toGuide` method checks for null, it will return null if the result of `datastore.get` is null. Therefore not checking works for this implementation.

If I remove the check in the `toGuide` method, then it will be necessary. Do you suggest that I do both changes?

I thought about implementing it like this from the beginning but decided not to go to this path. Since the datastore API returns null when not found, it makes sense to me that the helper does it as well.

Reply:
Yes, I would suggest both changes. First, it makes more sense to do the null check here because this is where you call `datastore.get`. Having the null check in `toGuide` doesn't really fit with the behavior of that method. Then, here in `get`, you have to go read the implementation of `toGuide` to know that you've handled the null case appropriately.

You're right that it's at least somewhat common to return null from a get method (namely from a Map). However, nulls tend to be a common source of bugs.

If you return null from this method, you have to remember to check for null from the Servlet, or else in the JavaScript, or else you risk another NPE. If you return an Optional<>, then that check becomes hard to forget.

The other issue with NPEs is that they are hard to debug. Imagine you forgot to check for null in the `doGet` method of the Servlet. The stack trace gives you a line number, and you would have to take that line number to inspect the code to figure out what happened. Imagine instead you throw an exception in `DatastoreGuides.get`, then it bubbles all the way up to your error handler. This way, the stack trace shows the error originating in `DatastoreGuides.get`, and it has a type like `NotFoundException`, so it's very clear what happened without looking at any code at all.

Reply:
Done in next commit!